### PR TITLE
Python 3 compatibility for generic.py

### PR DESCRIPTION
**Pull request by enedil**
`discrete_log` was renamed to more appropriate `integer_log`.

Performance improvements (better algorithms).

Cleaned up some messy hacks (like `hex(long(n))[2:-1]`).

### CODE REVIEW

**Comment by Pharisaeus** [comment]
The problem is that those changes completely break compatibility with Python 2 and I would really like to avoid this. The only reasonable solution I can think of is to make a different branch for the Python3 compatible version.
Can you change the base for this merge request to Py3 branch?

File: crypto_commons/generic.py [hunk 1]
```diff
@@ -61,8 +74,10 @@ def factorial(n):
     :param n: number
     :return: n!
     """
-    return multiply(list(long_range(1, n + 1)))
 
+    # on Python 3 this is 10 times faster than multiply(range(1,n+1))
```
Comment (Pharisaeus) [inline crypto_commons/generic.py]: comment is not necessary here

File: crypto_commons/generic.py [hunk 2]
```diff
@@ -92,10 +107,14 @@ def factor(n, limit=1000000):
     primes = get_primes(limit)
     for prime in primes:
         while n % prime == 0 and n > 2:
-            n /= prime
+            n //= prime
             factors.append(prime)
         if n < 2:
             break
+    else:
```
Comment (Pharisaeus) [inline crypto_commons/generic.py:120]: I don't get the point of this else condition.

File: crypto_commons/generic.py [hunk 3]
```diff
@@ -146,18 +165,23 @@ def long_range(start, stop, step=1):
         i += step
 
 
-def discrete_log(x, xi, limit=1000):
+def integer_log(x, xi, limit=1000):
```
Comment (Pharisaeus) [inline crypto_commons/generic.py]: https://en.wikipedia.org/wiki/Discrete_logarithm

File: crypto_commons/generic.py [hunk 4]
```diff
@@ -196,7 +220,11 @@ def xor_hex(t1, t2):
 
 def is_printable(data):
     import string
+    printable = set(string.printable)
```
Comment (Pharisaeus) [inline crypto_commons/generic.py]: Maybe simply `return len(set(data).intersection(set(string.printable))) > 0` ?

File: crypto_commons/generic.py [hunk 5]
```diff
@@ -1,22 +1,33 @@
+from builtins import range as long_range
```
Comment (Pharisaeus) [inline crypto_commons/generic.py]: This won't work for python 2 for large numbers (range is limited by maxint and creates the list in memory)

File: crypto_commons/xor/repeating_xor.py [hunk 6]
```diff
@@ -1,3 +1,4 @@
+from builtins import input, range
```
Comment (Pharisaeus) [inline crypto_commons/xor/repeating_xor.py]: requires `pip install future` but ok

**Comment by Pharisaeus** [comment]
Ok maybe I was a bit too hasty. Most changes seems ok. Please review the comments and I will be able to merge those.

File: crypto_commons/generic.py (hunk 4 above)
Comment (enedil) [reply crypto_commons/generic.py]: Or rather `len(set(data).difference(printable)) == 0`, your version will detect any string with a printable byte. Anyway, will be fixed.

Comment (Pharisaeus) [reply crypto_commons/generic.py]: This is what I was thinking about :)

**Comment by Pharisaeus** [comment]
Ping me when you think all changes you wanted are included in this merge request.

**Pull request merged**